### PR TITLE
Fix Transfer Crash

### LINKED ISSUE #671: Crash when transfering money to another account
Everytime i want to transfer money to another account it's just crashed

```
UnhandledException - unhandled exception: System.ArgumentOutOfRangeException: Exception of type 'System.ArgumentOutOfRangeException' was thrown. (Parameter 'ordinal')
Actual value was 13.
   at Microsoft.Data.Sqlite.SqliteDataRecord.GetSqliteType(Int32 ordinal)
   at Microsoft.Data.Sqlite.SqliteDataRecord.IsDBNull(Int32 ordinal)
   at NickvisionMoney.Shared.Models.AccountMetadata.LoadFromAccountFile(String path, String password) in /run/build/org.nickvision.money/NickvisionMoney.Shared/Models/AccountMetadata.cs:line 196
   at NickvisionMoney.Shared.Controllers.TransferDialogController.UpdateTransfer(String destPath, String destPassword, String amountString, String sourceConversionAmountString, String destConversionAmountString) in /run/build/org.nickvision.money/NickvisionMoney.Shared/Controllers/TransferDialogController.cs:line 156
   at NickvisionMoney.GNOME.Views.TransferDialog.Validate() in /run/build/org.nickvision.money/NickvisionMoney.GNOME/Views/TransferDialog.cs:line 135
   at NickvisionMoney.GNOME.Views.TransferDialog.<.ctor>b__21_1(Object sender, NotifySignalArgs e) in /run/build/org.nickvision.money/NickvisionMoney.GNOME/Views/TransferDialog.cs:line 61
   at GObject.Closure.InternalCallback(IntPtr closure, IntPtr returnValuePtr, UInt32 nParamValues, ValueData[] paramValuesData, IntPtr invocationHint, IntPtr userData)
```

### PR DESCRIPTION
**Pull request by nlogozzo**
Closes #671
Fixes https://github.com/NickvisionApps/Denaro/issues/671#issuecomment-1696738448

(Replaces #676)

### CODE REVIEW

**Comment by nlogozzo** [comment]
@JoseBritto Try this PR should be good

**Comment by JoseBritto** [comment]
Yea it should work. I'm not home rn so can't test. Will try once I get home. I added the startupCalled variable to prevent accidentally calling that method more than once in the future. The lock was there to prevent race conditions when accessing the variable. Its not part of the async code. As in if the method was to be called twice on two threads the method might actually end up running even if there was an if clause with a variable protecting it. 
Removing the second call works but I didn't find any other way to wait for initialization without adding arbitrary delay which looked bad to me :(

Didn't update the file from metadata because I'm not that good at sql😅 (also it will be updated anyway from somewhere else when the file is opened). 

Still think adding version info could be a good idea for the long term as it ensures we can change the way data is stored if needed in the future and not worry too much about breaking compatibility. 
On a similar note, if we abstract out the reading part from the upgrading the file part and upgrade only if needed, we can have a file open as readonly and if the user doesn't add any transactions, it wouldn't cause denaro to crash. I haven't tried to open a readonly file in denaro yet. Will try that too when I get home. 
Not sure if all of this would be worth the effort though haha. 

Let me know if you're interested in supporting opening readonly files. I'm not sure if it works as of now, but it is something I used to do with important files. But never tried with denaro.

**Comment by JoseBritto** [comment]
Still crashing:
```
UnhandledException - unhandled exception: Microsoft.Data.Sqlite.SqliteException (0x80004005): SQLite Error 19: 'UNIQUE constraint failed: transactions.id'.
   at Microsoft.Data.Sqlite.SqliteException.ThrowExceptionForRC(Int32 rc, sqlite3 db)
   at Microsoft.Data.Sqlite.SqliteDataReader.NextResult()
   at Microsoft.Data.Sqlite.SqliteCommand.ExecuteReader(CommandBehavior behavior)
   at Microsoft.Data.Sqlite.SqliteCommand.ExecuteReader()
   at Microsoft.Data.Sqlite.SqliteCommand.ExecuteNonQuery()
   at System.Data.Common.DbCommand.ExecuteNonQueryAsync(CancellationToken cancellationToken)
--- End of stack trace from previous location ---
   at NickvisionMoney.Shared.Models.Account.AddTransactionAsync(Transaction transaction) in /home/britto/Source/C#/Denaro/NickvisionMoney.Shared/Models/Account.cs:line 850
   at NickvisionMoney.Shared.Models.Account.ReceiveTransferAsync(Transfer transfer, String description) in /home/britto/Source/C#/Denaro/NickvisionMoney.Shared/Models/Account.cs:line 1249
   at NickvisionMoney.Shared.Controllers.AccountViewController.ReceiveTransferAsync(Transfer transfer) in /home/britto/Source/C#/Denaro/NickvisionMoney.Shared/Controllers/AccountViewController.cs:line 900
   at NickvisionMoney.Shared.Controllers.MainWindowController.OnTransferSent(Object sender, Transfer transfer) in /home/britto/Source/C#/Denaro/NickvisionMoney.Shared/Controllers/MainWindowController.cs:line 387
   at System.Threading.Tasks.Task.<>c.<ThrowAsync>b__128_0(Object state)
   at GLib.Internal.MainLoopSynchronizationContext.<>c__DisplayClass1_0.<Post>b__0()
   at GLib.Internal.MainLoopSynchronizationContext.<>c__DisplayClass3_0.<ScheduleAction>b__0()
```

This is probably due to a race condition. I had seen this same one before. Its painful to debug these type of bugs. This one mostly happens (for me) when the app is in debug mode and transferring to this particular [file](0x0.st/Hp8B.nmoney).

**Comment by JoseBritto** [comment]
Ha. Not a new bug. Seems like 500 isn't enough sometimes. This is the problem with hard coded delays. 1000 worked but it doesn't mean it will work everywhere tho.

File: NickvisionMoney.Shared/Controllers/MainWindowController.cs [hunk 1]
```diff
@@ -382,8 +382,8 @@ public void RemoveRecentAccount(RecentAccount recentAccount)
     private async void OnTransferSent(object? sender, Transfer transfer)
     {
         await AddAccountAsync(transfer.DestinationAccountPath, false, transfer.DestinationAccountPassword);
+        await Task.Delay(500);
```
Comment (fsobolev) [inline NickvisionMoney.Shared/Controllers/MainWindowController.cs:385]: I remember you once complained about using delays as workarounds making the app slower :D And @JoseBritto confirmed my fear that 500 can be not enough. Maybe add something like `await controller.EnsureStartup();` that will not return until startup is finished (similar to what I suggested in previous PR)?

**Review by DaPigGuy — CHANGES_REQUESTED** [review]

File: NickvisionMoney.Shared/Controllers/MainWindowController.cs (hunk 1 above)
Comment (nlogozzo) [reply NickvisionMoney.Shared/Controllers/MainWindowController.cs]: done in latest commit.

File: NickvisionMoney.Shared/Models/AccountMetadata.cs [hunk 2]
```diff
@@ -173,6 +173,56 @@ internal AccountMetadata(string name, AccountType accountType)
             database.Close();
             return null;
         }
+        // Update Metadata Table
+        try
```
[the remark this answers is not in the document]
Comment (nlogozzo) [reply NickvisionMoney.Shared/Models/AccountMetadata.cs:151]: done.

**Comment by nlogozzo** [comment]
@JoseBritto ok should be good now....I'm waiting for `IsOpened` to be `true` now instead of hardcoding a delay that way it can be as fast on small accounts and as slow on large accounts as it may need

**Comment by nlogozzo** [comment]
> Still think adding version info could be a good idea for the long term as it ensures we can change the way data is stored if needed in the future and not worry too much about breaking compatibility.

Well instead of versioning i'd like to keep it as it is now and just update the tables itself with the changes as they come...reduces the code too...I moved updating the metadata table to it's own function that both `AccountMetadata` and `Account` can call to avoid duplication as @DaPigGuy suggested

> On a similar note, if we abstract out the reading part from the upgrading the file part and upgrade only if needed, we can have a file open as readonly and if the user doesn't add any transactions, it wouldn't cause denaro to crash. I haven't tried to open a readonly file in denaro yet. Will try that too when I get home.
Not sure if all of this would be worth the effort though haha.

Yeah I don't really think we should support `readonly` files as I see no point as Denaro will have no use just besides viewing a list of transactions, which I mean just export to a PDF or something it'll be cleaner formatted anyway haha

**Comment by JoseBritto** [comment]
> If you mean checking "open file read-only" in the file chooser, it will not do anything, because we only use resulting GFile to get its path and then open the file using System.IO read-write anyway. And we always have read-write access in flatpak since we have home permission. And the reason for this is that I have no idea why but it seems SQL databases don't want to open through portal, I remember Denaro was crashing (this crash was fixed I think, it will just show that it can't open the file).

Oh wow. I didn't know that.

> Yeah I don't really think we should support readonly files as I see no point as Denaro will have no use just besides viewing a list of transactions, which I mean just export to a PDF or something it'll be cleaner formatted anyway haha

I meant exactly for viewing alone. Tbh, I prefer the UI of denaro lol. Also it allows me to filter by tags, see spending by groups, etc. That said, I don't use this kind of setup with denaro and it is probably a rare use case. Just wanted to throw it out there and see if that was something worth looking into. 

> Well instead of versioning i'd like to keep it as it is now and just update the tables itself with the changes as they come...reduces the code too...I moved updating the metadata table to it's own function that both AccountMetadata and Account can call to avoid duplication as @DaPigGuy suggested

Fair enough

**Comment by JoseBritto** [comment]
> @JoseBritto ok should be good now....I'm waiting for `IsOpened` to be `true` now instead of hardcoding a delay that way it can be as fast on small accounts and as slow on large accounts as it may need


That seem to have fixed it.

**Pull request merged**